stage1/iottymux/iottymux.go

```diff
@@ -0,0 +1,535 @@
+		if err != nil {
+			return err
+		}
+	}
+
+	// proxy stdin
+	if fifos[0] != nil && endpoints[0] != nil {
+		clients := make(chan *net.Conn)
+		go acceptConn(endpoints[0], clients, "stdin")
+		go muxInput(clients, fifos[0])
+	}
+
+	// proxy stdout
+	if fifos[1] != nil && endpoints[1] != nil {
+		localTargets := make(chan io.WriteCloser)
+		clients := make(chan *net.Conn)
+		lines := make(chan []byte)
+		go bufferLine(fifos[1], lines)
+		go acceptConn(endpoints[1], clients, "stdout")
+		go muxOutput("stdout", lines, clients, localTargets)
+		if logFile != nil {
+			localTargets <- logFile
+		}
+	}
+
+	// proxy stderr
+	if fifos[2] != nil && endpoints[2] != nil {
+		localTargets := make(chan io.WriteCloser)
+		clients := make(chan *net.Conn)
+		lines := make(chan []byte)
+		go bufferLine(fifos[2], lines)
+		go acceptConn(endpoints[2], clients, "stderr")
+		go muxOutput("stderr", lines, clients, localTargets)
+		if logFile != nil {
+			localTargets <- logFile
+		}
+	}
+
+	return <-c
+}
+
+// bufferLine buffers and queues a single line from a Reader to a multiplexer
+func bufferLine(r io.Reader, c chan []byte) {
+	rd := bufio.NewReader(r)
+	for {
+		lineOut, err := rd.ReadBytes('\n')
+		if err == nil {
+			c <- lineOut
+		}
+		// TODO(lucab): stop and signal caller on error
+	}
+}
+
+// acceptConn accepts a single client and queues it for further proxying
+func acceptConn(socket *net.Listener, c chan *net.Conn, stream string) {
+	for {
+		conn, err := (*socket).Accept()
+		if err == nil {
+			diag.Printf("Accepted new connection for %s", stream)
+			c <- &conn
+		}
+	}
+}
+
+// proxyIO performs bi- or mono-directional byte-by-byte forwarding
+func proxyIO(clients chan *net.Conn, tty *os.File, hasIn bool, hasOut bool) {
+	// TODO(lucab): this should probably become line-buffered and be muxed
+	// to logs as well. Still unsure.
+	for {
+		select {
+		case c := <-clients:
+			if hasIn {
+				go io.Copy(*c, tty)
+			}
+			if hasOut {
+				go io.Copy(tty, *c)
+			}
+		}
+		// TODO(lucab): stop and signal caller on error
+	}
+}
+
+// muxInput accepts remote client and multiplex input line from them
+func muxInput(clients chan *net.Conn, stdin *os.File) {
+	for {
+		select {
+		case c := <-clients:
+			go bufferInput(c, stdin)
+		}
+	}
+}
+
+// bufferInput buffers and write a single line from a remote client to the local app
+func bufferInput(conn *net.Conn, stdin *os.File) {
+	rd := bufio.NewReader(*conn)
+	for {
+		lineIn, err := rd.ReadBytes('\n')
+		if err != nil {
+			break
+		}
+		_, err = stdin.Write(lineIn)
+		if err != nil {
+			break
+		}
+	}
+}
+
+// muxOutput accepts remote clients and local log targets,
+// multiplexing output lines to them
+func muxOutput(streamLabel string, lines chan []byte, clients chan *net.Conn, targets chan io.WriteCloser) {
+	// TODO(lucab): this is a quick hack surviving too long. Rewrite all of it in a sane way.
+	var logs []io.WriteCloser
+	var alivel []bool
+	var conns []io.WriteCloser
+	var alivec []bool
```

Comment:
instead of maintaining `alivec` and `alivel` slices, you could remove them simply from the `logs` and `conns` slices in cases errors occur.